File: backend/internal/handlers/announcement.go
package handlers

import (
	"caiyun/internal/services"
	apiresponse "caiyun/pkg/response"
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
)

// AnnouncementHandler 公告处理器
type AnnouncementHandler struct {
	announcementService *services.AnnouncementService
}

// NewAnnouncementHandler 创建公告处理器
func NewAnnouncementHandler(announcementService *services.AnnouncementService) *AnnouncementHandler {
	return &AnnouncementHandler{
		announcementService: announcementService,
	}
}

// CreateAnnouncementRequest 创建公告请求
type CreateAnnouncementRequest struct {
	Title   string `json:"title" binding:"required"`
	Content string `json:"content" binding:"required"`
	IsPopup bool   `json:"is_popup"`
	IsTop   bool   `json:"is_top"`
}

// CreateAnnouncement 创建公告
func (h *AnnouncementHandler) CreateAnnouncement(c *gin.Context) {
	var req CreateAnnouncementRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		respondError(c, http.StatusBadRequest, "请求参数错误: "+err.Error())
		return
	}

	announcement, err := h.announcementService.CreateAnnouncement(req.Title, req.Content, req.IsPopup, req.IsTop)
	if err != nil {
		respondInternalServer(c)
		return
	}

	apiresponse.SuccessWithMessage(c, "创建成功", gin.H{"announcement": announcement})
}

// UpdateAnnouncementRequest 更新公告请求
type UpdateAnnouncementRequest struct {
	Title       string `json:"title" binding:"required"`
	Content     string `json:"content" binding:"required"`
	IsPopup     bool   `json:"is_popup"`
	IsTop       bool   `json:"is_top"`
	IsPublished bool   `json:"is_published"`
}

// UpdateAnnouncement 更新公告
func (h *AnnouncementHandler) UpdateAnnouncement(c *gin.Context) {
	idStr := c.Param("id")
	id, err := strconv.ParseUint(idStr, 10, 32)
	if err != nil {
		respondError(c, http.StatusBadRequest, "无效的ID")
		return
	}

	var req UpdateAnnouncementRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		respondError(c, http.StatusBadRequest, "请求参数错误: "+err.Error())
		return
	}

	announcement, err := h.announcementService.UpdateAnnouncement(uint(id), req.Title, req.Content, req.IsPopup, req.IsTop, req.IsPublished)
	if err != nil {
		respondInternalServer(c)
		return
	}

	apiresponse.SuccessWithMessage(c, "更新成功", gin.H{"announcement": announcement})
}

// DeleteAnnouncement 删除公告
func (h *AnnouncementHandler) DeleteAnnouncement(c *gin.Context) {
	idStr := c.Param("id")
	id, err := strconv.ParseUint(idStr, 10, 32)
	if err != nil {
		respondError(c, http.StatusBadRequest, "无效的ID")
		return
	}

	if err := h.announcementService.DeleteAnnouncement(uint(id)); err != nil {
		respondInternalServer(c)
		return
	}

	apiresponse.Message(c, "删除成功")
}

// GetAnnouncement 获取公告详情
func (h *AnnouncementHandler) GetAnnouncement(c *gin.Context) {
	idStr := c.Param("id")
	id, err := strconv.ParseUint(idStr, 10, 32)
	if err != nil {
		respondError(c, http.StatusBadRequest, "无效的ID")
		return
	}

	announcement, err := h.announcementService.GetAnnouncement(uint(id))
	if err != nil {
		respondError(c, http.StatusNotFound, "公告不存在")
		return
	}

	apiresponse.Success(c, gin.H{
		"announcement": announcement,
	})
}

// GetAllAnnouncements 获取所有公告
func (h *AnnouncementHandler) GetAllAnnouncements(c *gin.Context) {
	announcements, err := h.announcementService.GetAllAnnouncements()
	if err != nil {
		respondInternalServer(c)
		return
	}

	apiresponse.Success(c, gin.H{
		"announcements": announcements,
		"total":         len(announcements),
	})
}

// GetPublishedAnnouncements 获取已发布的公告
func (h *AnnouncementHandler) GetPublishedAnnouncements(c *gin.Context) {
	announcements, err := h.announcementService.GetPublishedAnnouncements()
	if err != nil {
		respondInternalServer(c)
		return
	}

	apiresponse.Success(c, gin.H{
		"announcements": announcements,
		"total":         len(announcements),
	})
}

// GetPopupAnnouncement 获取弹窗公告列表
func (h *AnnouncementHandler) GetPopupAnnouncement(c *gin.Context) {
	announcements, err := h.announcementService.GetPopupAnnouncements()
	if err != nil || len(announcements) == 0 {
		apiresponse.Success(c, gin.H{
			"has_popup":     false,
			"announcements": []interface{}{},
		})
		return
	}

	apiresponse.Success(c, gin.H{
		"has_popup":     true,
		"announcements": announcements,
	})
}

File: frontend/src/api/announcement.ts
import { http } from './axios'
import { unwrapApiData, type ApiResponse } from './response'

export interface Announcement {
  id: number
  title: string
  content: string
  is_popup: boolean
  is_top: boolean
  is_published: boolean
  popup_count: number
  created_at: string
  updated_at: string
}

export interface CreateAnnouncementRequest {
  title: string
  content: string
  is_popup: boolean
  is_top: boolean
}

export interface UpdateAnnouncementRequest {
  title: string
  content: string
  is_popup: boolean
  is_top: boolean
  is_published: boolean
}

export interface AnnouncementListResponse {
  announcements: Announcement[]
  total: number
}

export interface AnnouncementDetailResponse {
  announcement: Announcement | null
}

export interface PopupAnnouncementResponse {
  has_popup: boolean
  announcements: Announcement[]
}

const emptyList: AnnouncementListResponse = {
  announcements: [],
  total: 0
}

// 获取已发布的公告列表
export function getAnnouncements(): Promise<AnnouncementListResponse> {
  return http.get<AnnouncementListResponse | ApiResponse<AnnouncementListResponse>>('/api/announcements')
    .then((res) => unwrapApiData(res, emptyList))
}

// 获取弹窗公告
export function getPopupAnnouncement(): Promise<PopupAnnouncementResponse> {
  const fallback: PopupAnnouncementResponse = { has_popup: false, announcements: [] }
  return http.get<PopupAnnouncementResponse | ApiResponse<PopupAnnouncementResponse>>('/api/announcements/popup')
    .then((res) => unwrapApiData(res, fallback))
}

// 管理员：获取所有公告
export function getAllAnnouncements(): Promise<AnnouncementListResponse> {
  return http.get<AnnouncementListResponse | ApiResponse<AnnouncementListResponse>>('/api/admin/announcements')
    .then((res) => unwrapApiData(res, emptyList))
}

// 管理员：创建公告
export function createAnnouncement(data: CreateAnnouncementRequest): Promise<AnnouncementDetailResponse> {
  const fallback: AnnouncementDetailResponse = { announcement: null }
  return http.post<AnnouncementDetailResponse | ApiResponse<AnnouncementDetailResponse>>('/api/admin/announcements', data)
    .then((res) => unwrapApiData(res, fallback))
}

// 管理员：更新公告
export function updateAnnouncement(id: number, data: UpdateAnnouncementRequest): Promise<AnnouncementDetailResponse> {
  const fallback: AnnouncementDetailResponse = { announcement: null }
  return http.put<AnnouncementDetailResponse | ApiResponse<AnnouncementDetailResponse>>(`/api/admin/announcements/${id}`, data)
    .then((res) => unwrapApiData(res, fallback))
}

// 管理员：删除公告
export function deleteAnnouncement(id: number): Promise<{ message: string }> {
  return http.delete<{ message: string }>(`/api/admin/announcements/${id}`)
}

// 管理员：获取公告详情
export function getAnnouncementDetail(id: number): Promise<AnnouncementDetailResponse> {
  const fallback: AnnouncementDetailResponse = { announcement: null }
  return http.get<AnnouncementDetailResponse | ApiResponse<AnnouncementDetailResponse>>(`/api/admin/announcements/${id}`)
    .then((res) => unwrapApiData(res, fallback))
}
